### ISSUE TITLE
Automatic splitting of batch during model feed-forwarding to avoid memory errors caused by the too big batch_size

### ISSUE THREAD

**Issue by eaedk** [pytorch/pytorch#52036]
## 🚀 Feature
Ensure that models can according to the memory available on the GPU during a feed-forward, automatically split the sent batch into several small mini_batches to avoid the memory errors. 

## Motivation
So that during the training, users no longer have to reduce batch_size (or image size in case of image). The training will just take time but any users could train models any batch_size without check manually the good size for mini_batch.
During inference also, user could just pass a big batch_size and the framework will automatically define the best number of sample to passe to don't have memory errors anymore.

## Pitch

what I want is that whatever the size of the batch that we set, the GPU can split it into several small batches to process so as to avoid memory errors due to the size of the batch being too large.

**Comment by VitalyFedyunin** [comment]
There is no guarantee that models memory usage function is only depends on batch size (as it can also depend on inputs).

**Comment by eaedk** [comment]
yeah I know that, but I think in many cases there are enough space on gpu to feed the model with at least 2-4 inputs so and automatic batch splitting will avoid users to make many combinations and just train models whatever batch_size they set.
